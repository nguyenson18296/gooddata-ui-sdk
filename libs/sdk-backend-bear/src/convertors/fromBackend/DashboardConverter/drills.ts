// (C) 2019-2021 GoodData Corporation

import { GdcVisualizationWidget, GdcKpi } from "@gooddata/api-model-bear";
import { uriRef, idRef } from "@gooddata/sdk-model";
import {
    IDrillToLegacyDashboard,
    DrillDefinition,
    IDrillToDashboard,
    IDrillToInsight,
    IDrillToCustomUrl,
    IDrillToAttributeUrl,
    UnexpectedError,
    DrillOrigin,
} from "@gooddata/sdk-backend-spi";

export const convertKpiDrill = (kpi: GdcKpi.IWrappedKPI): IDrillToLegacyDashboard => {
    const { drillTo: { projectDashboard, projectDashboardTab } = {}, metric } = kpi.kpi.content;

    const drillDefinition: IDrillToLegacyDashboard = {
        type: "drillToLegacyDashboard",
        origin: {
            type: "drillFromMeasure",
            measure: uriRef(metric),
        },
        target: uriRef(projectDashboard!),
        tab: projectDashboardTab!,
        transition: "in-place",
    };

    return drillDefinition;
};

export const convertDrillOrigin = (from: GdcVisualizationWidget.DrillFromType): DrillOrigin => {
    if (GdcVisualizationWidget.isDrillFromMeasure(from)) {
        return {
            type: "drillFromMeasure",
            measure: from.drillFromMeasure,
        };
    } else if (GdcVisualizationWidget.isDrillFromAttribute(from)) {
        return {
            type: "drillFromAttribute",
            attribute: from.drillFromAttribute,
        };
    } else {
        throw new UnexpectedError("Unable to convert unknown drill origin!");
    }
};

export const convertVisualizationWidgetDrill = (
    drill: GdcVisualizationWidget.IDrillDefinition,
): DrillDefinition => {
    if (GdcVisualizationWidget.isDrillToDashboard(drill)) {
        const {
            drillToDashboard: { toDashboard, target, from },
        } = drill;
        const drillDefinition: IDrillToDashboard = {
            type: "drillToDashboard",
            origin: convertDrillOrigin(from),
            target: toDashboard !== undefined ? idRef(toDashboard) : undefined,
            transition: target,
        };

        return drillDefinition;
    } else if (GdcVisualizationWidget.isDrillToVisualization(drill)) {
        const {
            drillToVisualization: { toVisualization, target, from },
        } = drill;
        const drillDefinition: IDrillToInsight = {
            type: "drillToInsight",
            origin: convertDrillOrigin(from),
            target: toVisualization,
            transition: target,
        };

        return drillDefinition;
    } else if (GdcVisualizationWidget.isDrillToCustomUrl(drill)) {
        const {
            drillToCustomUrl: { target, customUrl, from },
        } = drill;
        const drillDefinition: IDrillToCustomUrl = {
            type: "drillToCustomUrl",
            origin: convertDrillOrigin(from),
            target: {
                url: customUrl,
            },
            transition: target,
        };

        return drillDefinition;
    } else if (GdcVisualizationWidget.isDrillToAttributeUrl(drill)) {
        const {
            drillToAttributeUrl: { drillToAttributeDisplayForm, insightAttributeDisplayForm, target, from },
        } = drill;
        const drillDefinition: IDrillToAttributeUrl = {
            type: "drillToAttributeUrl",
            origin: convertDrillOrigin(from),
            target: {
                displayForm: insightAttributeDisplayForm,
                hyperlinkDisplayForm: drillToAttributeDisplayForm,
            },
            transition: target,
        };

        return drillDefinition;
    }

    throw new UnexpectedError("Unable to convert unknown drill!");
};
