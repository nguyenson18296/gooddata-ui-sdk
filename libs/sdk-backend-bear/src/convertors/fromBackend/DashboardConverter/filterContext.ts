// (C) 2019-2021 GoodData Corporation
import { GdcFilterContext } from "@gooddata/api-model-bear";
import { uriRef } from "@gooddata/sdk-model";
import {
    IFilterContext,
    FilterContextItem,
    IDashboardDateFilter,
    ITempFilterContext,
    isDashboardDateFilter,
    IFilterContextDefinition,
    IDashboardAttributeFilterParent,
} from "@gooddata/sdk-backend-spi";

function isNotTemporaryAllTimeDateFilter(filter: FilterContextItem): boolean {
    if (isDashboardDateFilter(filter)) {
        return filter.dateFilter.from !== undefined || filter.dateFilter.to !== undefined;
    }

    return true;
}

// Remove the temporary "All Time" date filter from filter context when exporting the dashboard
export const sanitizeExportFilterContext = (
    exportFilterContext: IFilterContext | ITempFilterContext,
): IFilterContext | ITempFilterContext => {
    return {
        ...exportFilterContext,
        filters: exportFilterContext.filters.filter(isNotTemporaryAllTimeDateFilter),
    };
};

export const convertFilterContextItem = (
    filterContextItem: GdcFilterContext.FilterContextItem,
): FilterContextItem => {
    if (GdcFilterContext.isAttributeFilter(filterContextItem)) {
        const {
            attributeFilter: {
                attributeElements,
                displayForm,
                negativeSelection,
                localIdentifier,
                filterElementsBy = [],
            },
        } = filterContextItem;

        const convertedFilterElementsBy: IDashboardAttributeFilterParent[] = filterElementsBy.map(
            (filterBy) => {
                return {
                    filterLocalIdentifier: filterBy.filterLocalIdentifier,
                    over: {
                        attributes: filterBy.over.attributes.map(uriRef),
                    },
                };
            },
        );

        return {
            attributeFilter: {
                attributeElements: { uris: attributeElements },
                displayForm: uriRef(displayForm),
                negativeSelection,
                localIdentifier,
                filterElementsBy: convertedFilterElementsBy,
            },
        };
    }
    const {
        dateFilter: { granularity, type, attribute, dataSet, from, to },
    } = filterContextItem;
    const convertedFilterContextItem: IDashboardDateFilter = {
        dateFilter: {
            granularity,
            type,
            from,
            to,
        },
    };
    if (attribute) {
        convertedFilterContextItem.dateFilter.attribute = uriRef(attribute);
    }
    if (dataSet) {
        convertedFilterContextItem.dateFilter.dataSet = uriRef(dataSet);
    }

    return convertedFilterContextItem;
};

export const convertFilterContext = (
    filterContext: GdcFilterContext.IWrappedFilterContext,
): IFilterContext | IFilterContextDefinition => {
    const {
        filterContext: {
            content: { filters },
            meta: { identifier, uri, summary, title },
        },
    } = filterContext;

    return {
        description: summary!,
        ...(uri
            ? {
                  identifier,
                  uri,
                  ref: uriRef(uri),
              }
            : {}),
        title,
        filters: filters.map(convertFilterContextItem),
    };
};

export const convertTempFilterContext = (
    filterContext: GdcFilterContext.IWrappedTempFilterContext,
): ITempFilterContext => {
    const {
        tempFilterContext: { created, filters, uri },
    } = filterContext;

    return {
        uri,
        ref: uriRef(uri),
        filters: filters.map(convertFilterContextItem),
        created,
    };
};
