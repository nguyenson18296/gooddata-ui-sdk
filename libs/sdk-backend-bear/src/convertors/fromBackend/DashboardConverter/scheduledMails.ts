// (C) 2019-2020 GoodData Corporation
import { GdcScheduledMail } from "@gooddata/api-model-bear";
import {
    IScheduledMail,
    IScheduledMailDefinition,
    NotSupported,
    ScheduledMailAttachment,
} from "@gooddata/sdk-backend-spi";
import { uriRef } from "@gooddata/sdk-model";

export const convertScheduledMailAttachment = (
    scheduledMailAttachment: GdcScheduledMail.ScheduledMailAttachment,
): ScheduledMailAttachment => {
    if (!GdcScheduledMail.isKpiDashboardAttachment(scheduledMailAttachment)) {
        throw new NotSupported(
            "Cannot convert attachment - only dashboard attachments are currently supported.",
        );
    }

    const {
        kpiDashboardAttachment: { format, uri, filterContext },
    } = scheduledMailAttachment;

    return {
        dashboard: uriRef(uri),
        format,
        filterContext: filterContext ? uriRef(filterContext) : undefined,
    };
};

export const convertScheduledMail = (
    scheduledMail: GdcScheduledMail.IWrappedScheduledMail,
): IScheduledMail | IScheduledMailDefinition => {
    const {
        scheduledMail: {
            content: { attachments, body, subject, to, when, bcc, lastSuccessfull, unsubscribed },
            meta: { uri, identifier, title, summary, unlisted },
        },
    } = scheduledMail;

    return {
        title,
        description: summary!,
        ...(uri
            ? {
                  ref: uriRef(uri),
                  identifier,
                  uri,
              }
            : {}),
        body,
        subject,
        to,
        when: {
            startDate: when.startDate,
            endDate: when.endDate,
            timeZone: when.timeZone,
            recurrence: when.recurrency,
        },
        bcc,
        lastSuccessful: lastSuccessfull,
        unsubscribed,
        attachments: attachments.map(convertScheduledMailAttachment),
        unlisted: !!unlisted,
    };
};
