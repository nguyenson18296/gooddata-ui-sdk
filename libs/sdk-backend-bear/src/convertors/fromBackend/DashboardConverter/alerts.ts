// (C) 2019-2020 GoodData Corporation
import { GdcMetadata, GdcFilterContext } from "@gooddata/api-model-bear";
import { IWidgetAlert, IWidgetAlertDefinition } from "@gooddata/sdk-backend-spi";
import { uriRef } from "@gooddata/sdk-model";
import { convertFilterContext } from "./filterContext";

export const convertAlert = (
    alert: GdcMetadata.IWrappedKpiAlert,
    filterContext?: GdcFilterContext.IWrappedFilterContext,
): IWidgetAlert | IWidgetAlertDefinition => {
    const {
        kpiAlert: {
            content: { dashboard, isTriggered, kpi, threshold, whenTriggered },
            meta: { uri, identifier, title, summary },
        },
    } = alert;
    return {
        title,
        description: summary!,
        ...(uri
            ? {
                  ref: uriRef(uri),
                  identifier,
                  uri,
              }
            : {}),
        dashboard: uriRef(dashboard),
        widget: uriRef(kpi),
        threshold,
        whenTriggered,
        isTriggered,
        filterContext: filterContext && convertFilterContext(filterContext),
    };
};
