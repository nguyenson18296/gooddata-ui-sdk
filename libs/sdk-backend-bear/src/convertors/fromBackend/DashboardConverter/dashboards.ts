// (C) 2019-2021 GoodData Corporation

import {
    GdcMetadata,
    GdcDashboard,
    GdcFilterContext,
    GdcVisualizationWidget,
    GdcKpi,
    GdcVisualizationObject,
    GdcVisualizationClass,
} from "@gooddata/api-model-bear";
import { uriRef } from "@gooddata/sdk-model";
import keyBy from "lodash/keyBy";
import {
    IListedDashboard,
    IDashboard,
    IFilterContext,
    IDashboardDateFilterAddedPresets,
    ITempFilterContext,
    isWidget,
    IDashboardDateFilterConfig,
} from "@gooddata/sdk-backend-spi";
import { sanitizeExportFilterContext, convertFilterContext, convertTempFilterContext } from "./filterContext";
import { convertLayout, createImplicitDashboardLayout } from "./layout";
import { DashboardDependency, BearDashboardDependency } from "./types";
import { convertVisualizationWidget, convertKpi } from "./widget";

export const convertListedDashboard = (dashboardLink: GdcMetadata.IObjectLink): IListedDashboard => ({
    ref: uriRef(dashboardLink.link),
    identifier: dashboardLink.identifier!,
    uri: dashboardLink.link,
    title: dashboardLink.title!,
    description: dashboardLink.summary!,
    updated: dashboardLink.updated!,
    created: dashboardLink.created!,
    // filter takes care of multiple spaces and also the base scenario ("" ~> [])
    tags: dashboardLink.tags?.split(" ").filter(Boolean),
});

const convertDateFilterConfigAddedPresets = (
    addPresets: GdcDashboard.IDashboardDateFilterAddedPresets,
): IDashboardDateFilterAddedPresets => {
    const { absolutePresets = [], relativePresets = [] } = addPresets;
    return {
        absolutePresets: absolutePresets.map((preset) => ({ ...preset, type: "absolutePreset" })),
        relativePresets: relativePresets.map((preset) => ({ ...preset, type: "relativePreset" })),
    };
};

export const convertDashboardDateFilterConfig = (
    dateFilterConfig: GdcDashboard.IDashboardDateFilterConfig,
): IDashboardDateFilterConfig => {
    const { filterName, mode, addPresets, hideGranularities, hideOptions } = dateFilterConfig;

    return {
        filterName,
        mode,
        addPresets: addPresets && convertDateFilterConfigAddedPresets(addPresets),
        hideGranularities,
        hideOptions,
    };
};

export const convertDashboard = (
    dashboard: GdcDashboard.IWrappedAnalyticalDashboard,
    dependencies: BearDashboardDependency[],
    visualizationClasses: GdcVisualizationClass.IVisualizationClassWrapped[] = [],
    exportFilterContextUri?: string,
): IDashboard => {
    const {
        meta: { summary, created, updated, identifier, uri, title, locked, tags },
        content: { layout, filterContext, dateFilterConfig, widgets: widgetsUris },
    } = dashboard.analyticalDashboard;

    const sdkDependencies = dependencies
        // Filter out visualization objects - we only need them to create implicit layout
        .filter((d) => !GdcVisualizationObject.isVisualization(d))
        .map(convertDashboardDependency);
    const unsortedWidgets = sdkDependencies.filter(isWidget);

    // To preserve the logic of createImplicitDashboardLayout, we must preserve the order of the widgets
    const widgetByUri = keyBy(unsortedWidgets, "uri");
    const widgets = widgetsUris.map((widgetUri) => widgetByUri[widgetUri]);

    const filterContextOrExportFilterContext = sdkDependencies.find((dep) =>
        exportFilterContextUri ? dep.uri === exportFilterContextUri : dep.uri === filterContext,
    ) as IFilterContext | ITempFilterContext | undefined;

    return {
        title,
        description: summary!,

        identifier: identifier!,
        uri: uri!,
        ref: uriRef(uri!),

        created: created!,
        updated: updated!,
        isLocked: !!locked,

        dateFilterConfig: dateFilterConfig && convertDashboardDateFilterConfig(dateFilterConfig),

        filterContext:
            exportFilterContextUri && filterContextOrExportFilterContext
                ? sanitizeExportFilterContext(filterContextOrExportFilterContext)
                : filterContextOrExportFilterContext,
        layout: layout
            ? convertLayout(layout, widgets)
            : createImplicitDashboardLayout(widgets, dependencies, visualizationClasses),

        // filter takes care of multiple spaces and also the base scenario ("" ~> [])
        tags: tags?.split(" ").filter((t) => t),
    };
};

const convertDashboardDependency = (dependency: BearDashboardDependency): DashboardDependency => {
    if (GdcVisualizationWidget.isWrappedVisualizationWidget(dependency)) {
        return convertVisualizationWidget(dependency);
    } else if (GdcKpi.isWrappedKpi(dependency)) {
        return convertKpi(dependency);
    } else if (GdcFilterContext.isWrappedFilterContext(dependency)) {
        return convertFilterContext(dependency) as IFilterContext;
    } else if (GdcFilterContext.isWrappedTempFilterContext(dependency)) {
        return convertTempFilterContext(dependency);
    }

    throw new Error(`No converter for the dashboard dependency!`);
};
