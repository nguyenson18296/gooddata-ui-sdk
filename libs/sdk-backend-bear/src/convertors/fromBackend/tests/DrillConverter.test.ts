// (C) 2021 GoodData Corporation

import { convertDrillOrigin, convertVisualizationWidgetDrill } from "../DashboardConverter";
import {
    drillFromAttribute,
    drillFromMeasure,
    drillToDashboardWithDrillFromMeasure,
    drillToDashboardWithDrillFromAttribute,
} from "./DrillConverter.fixtures";

describe("convert drill", () => {
    describe("origin", () => {
        it("should convert drill from measure", () => {
            const convertedDrill = convertDrillOrigin(drillFromMeasure);
            expect(convertedDrill).toMatchSnapshot();
        });

        it("should convert drill from attribute", () => {
            const convertedDrill = convertDrillOrigin(drillFromAttribute);
            expect(convertedDrill).toMatchSnapshot();
        });
    });

    describe("definition", () => {
        it("should convert drill to dashboard with drill from measure", () => {
            const convertedDrill = convertVisualizationWidgetDrill(drillToDashboardWithDrillFromMeasure);
            expect(convertedDrill).toMatchSnapshot();
        });

        it("should convert drill to dashboard with drill from attribute", () => {
            const convertedDrill = convertVisualizationWidgetDrill(drillToDashboardWithDrillFromAttribute);
            expect(convertedDrill).toMatchSnapshot();
        });
    });
});
