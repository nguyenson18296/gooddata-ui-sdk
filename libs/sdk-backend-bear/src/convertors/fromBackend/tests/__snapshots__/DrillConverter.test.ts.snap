// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert drill definition should convert drill to dashboard with drill from attribute 1`] = `
Object {
  "origin": Object {
    "attribute": Object {
      "localIdentifier": "attributeLocalIdentifier",
    },
    "type": "drillFromAttribute",
  },
  "target": Object {
    "identifier": "dashboardId",
  },
  "transition": "in-place",
  "type": "drillToDashboard",
}
`;

exports[`convert drill definition should convert drill to dashboard with drill from measure 1`] = `
Object {
  "origin": Object {
    "measure": Object {
      "localIdentifier": "measureLocalIdentifier",
    },
    "type": "drillFromMeasure",
  },
  "target": Object {
    "identifier": "dashboardId",
  },
  "transition": "in-place",
  "type": "drillToDashboard",
}
`;

exports[`convert drill origin should convert drill from attribute 1`] = `
Object {
  "attribute": Object {
    "localIdentifier": "attributeLocalIdentifier",
  },
  "type": "drillFromAttribute",
}
`;

exports[`convert drill origin should convert drill from measure 1`] = `
Object {
  "measure": Object {
    "localIdentifier": "measureLocalIdentifier",
  },
  "type": "drillFromMeasure",
}
`;
